src/chainrules.jl

```diff
@@ -0,0 +1,160 @@
+        return NO_FIELDS, NO_FIELDS, Zero(), Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix; dims=2
+)
+    P = Distances.pairwise(d, X; dims=dims)
+    function pairwise_pullback(::AbstractMatrix)
+        return NO_FIELDS, NO_FIELDS, Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.colwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix
+)
+    C = Distances.colwise(d, X, Y)
+    function colwise_pullback(::AbstractVector)
+        return NO_FIELDS, NO_FIELDS, Zero(), Zero()
+    end
+    return C, colwise_pullback
+end
+
+## Reverse Rules DotProduct
+
+function ChainRulesCore.rrule(dist::DotProduct, x::AbstractVector, y::AbstractVector)
+    d = dist(x, y)
+    function evaluate_pullback(Δ::Any)
+        return NO_FIELDS, Δ .* y, Δ .* x
+    end
+    return d, evaluate_pullback
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.pairwise),
+    d::DotProduct,
+    X::AbstractMatrix,
+    Y::AbstractMatrix;
+    dims=2,
+)
+    P = Distances.pairwise(d, X, Y; dims=dims)
+    function pairwise_pullback_cols(Δ::AbstractMatrix)
+        if dims == 1
+            return NO_FIELDS, NO_FIELDS, Δ * Y, Δ' * X
+        else
+            return NO_FIELDS, NO_FIELDS, Y * Δ', X * Δ
+        end
+    end
+    return P, pairwise_pullback_cols
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.pairwise), d::DotProduct, X::AbstractMatrix; dims=2
+)
+    P = Distances.pairwise(d, X; dims=dims)
+    function pairwise_pullback_cols(Δ::AbstractMatrix)
+        if dims == 1
+            return NO_FIELDS, NO_FIELDS, 2 * Δ * X
+        else
+            return NO_FIELDS, NO_FIELDS, 2 * X * Δ
+        end
+    end
+    return P, pairwise_pullback_cols
+end
+
+function ChainRulesCore.rrule(
+    ::typeof(Distances.colwise), d::DotProduct, X::AbstractMatrix, Y::AbstractMatrix
+)
+    C = Distances.colwise(d, X, Y)
+    function colwise_pullback(Δ::AbstractVector)
+        return NO_FIELDS, NO_FIELDS, Δ' .* Y, Δ' .* X
+    end
+    return C, colwise_pullback
+end
+
+## Reverse Rules Sinus
+
+function ChainRulesCore.rrule(s::Sinus, x::AbstractVector, y::AbstractVector)
+    d = x - y
+    sind = sinpi.(d)
+    abs2_sind_r = abs2.(sind) ./ s.r
+    val = sum(abs2_sind_r)
+    gradx = twoπ .* cospi.(d) .* sind ./ (s.r .^ 2)
+    function evaluate_pullback(Δ::Any)
+        return (r=-2Δ .* abs2_sind_r,), Δ * gradx, -Δ * gradx
+    end
+    return val, evaluate_pullback
+end
+
+## Reverse Rulse SqMahalanobis
+
+function ChainRulesCore.rrule(
+    dist::Distances.SqMahalanobis, a::AbstractVector, b::AbstractVector
+)
+    d = dist(a, b)
+    function SqMahalanobis_pullback(Δ::Real)
+        B_Bᵀ = dist.qmat + transpose(dist.qmat)
+        a_b = a - b
+        δa = @thunk((B_Bᵀ * a_b) * Δ)
+        return (qmat=(a_b * a_b') * Δ,), δa, @thunk(-δa)
```

Comment:
I think there might not be any benefit from putting the gradient of `b` in a thunk since the construction of the closure might be too expensive. Moreover, it seems one can safely assume that `qmat` is symmetric since the definition of `pairwise!` in Distances uses this property (see https://github.com/JuliaStats/Distances.jl/issues/174). After reading https://juliadiff.org/ChainRulesCore.jl/dev/gradient_accumulation.html, my suggestion would be:
```suggestion
        a_b = a - b
        ∂qmat = InplaceableThunk(
            @thunk((a_b * a_b') * Δ),
            X̄ -> mul!(X̄, a_b, a_b', true, Δ),
        )
        ∂a = InplaceableThunk(
            @thunk((2 * Δ) * dist.qmat * a_b),
            X̄ -> mul!(X̄, dist.qmat, a_b, true, 2 * Δ),
        )
        ∂b = InplaceableThunk(
            @thunk((-2 * Δ) * dist.qmat * a_b),
            X̄ -> mul!(X̄, dist.qmat, a_b, true, -2 * Δ),
        )
        return Composite{typeof(dist)}(qmat = ∂qmat), ∂a, ∂b
```